front/FrontService.cpp

```diff
@@ -19,5 +19,478 @@
+                                            bytesConstRef _data,
+                                            uint32_t _timeout,
+                                            CallbackFunc _callbackFunc) {
+
+  try {
+    std::string uuid;
+
+    if (_callbackFunc) {
+      uuid = boost::uuids::to_string(boost::uuids::random_generator()());
+
+      auto callback = std::make_shared<Callback>();
+      callback->callbackFunc = _callbackFunc;
+
+      if (_timeout > 0) {
+        // create new timer to handle timeout
+        auto timeoutHandler = std::make_shared<boost::asio::deadline_timer>(
+            *m_ioService, boost::posix_time::milliseconds(_timeout));
+
+        callback->timeoutHandler = timeoutHandler;
+        // callback->startTime = utcSteadyTime();
+        timeoutHandler->async_wait([frontServicePtr = shared_from_this(),
+                                    uuid](const boost::system::error_code &e) {
+          frontServicePtr->onMessageTimeout(e, uuid);
+        });
+      }
+
+      addCallback(uuid, callback);
+
+      FRONT_LOG(DEBUG) << LOG_BADGE("asyncSendMessageByNodeID")
+                       << LOG_KV("uuid", uuid) << LOG_KV("groupID", m_groupID)
+                       << LOG_KV("nodeID", m_nodeID->hex())
+                       << LOG_KV("timeout", _timeout);
+
+    } // if (_callback)
+
+    onSendMessage(_moduleID, _nodeID, uuid, _data);
+
+  } catch (std::exception &e) {
+    FRONT_LOG(ERROR) << LOG_BADGE("asyncSendMessageByNodeID")
+                     << LOG_KV("what", boost::diagnostic_information(e));
+  }
+}
+
+/**
+ * @brief: send message to multiple nodes
+ * @param _moduleID: moduleID
+ * @param _nodeIDs: the receiver nodeIDs
+ * @param _data: send data
+ * @return void
+ */
+void FrontService::asyncSendMessageByNodeIDs(int _moduleID,
+                                             const crypto::NodeIDs &_nodeIDs,
+                                             bytesConstRef _data) {
+  for (const auto &_nodeID : _nodeIDs) {
+    asyncSendMessageByNodeID(_moduleID, _nodeID, _data, 0, CallbackFunc());
+  }
+}
+
+/**
+ * @brief: send broadcast message
+ * @param _moduleID: moduleID
+ * @param _data:  send data
+ * @return void
+ */
+void FrontService::asyncMulticastMessage(int _moduleID, bytesConstRef _data) {
+
+  auto message = messageFactory()->buildMessage();
+  message->setModuleID(_moduleID);
+  message->setPayload(_data);
+
+  auto buffer = std::make_shared<bytes>();
+  message->encode(*buffer.get());
+
+  m_gatewayInterface->asyncMulticastMessage(
+      m_groupID, bytesConstRef(buffer->data(), buffer->size()));
+}
+
+/**
+ * @brief: register the node change callback
+ * @param _moduleID: moduleID
+ * @param _notifier: callback
+ * @return void
+ */
+void FrontService::registerNodeStatusNotifier(int _moduleID,
+                                              NodeStatusNotifier _notifier) {
+  m_mapNodeStatusNotifier[_moduleID] = _notifier;
```

Comment:
Should it be protected by a lock?

Reply:
the Register interface is called during initialization.